### PR TITLE
Code smells following Java 17 migration

### PR DESCRIPTION
**Pull request by nquinquenel**

### CODE REVIEW

File: src/main/java/org/sonarlint/intellij/core/ProjectBindingManager.java [hunk 1]
```diff
@@ -244,40 +244,20 @@ public Set<ProjectKeyAndBranch> getUniqueProjectKeysAndBranchesPairs() {
...
+      @Override
+      public boolean equals(Object o) {
+        if (this == o) {
+          return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+          return false;
+        }
+        ProjectKeyAndBranch that = (ProjectKeyAndBranch) o;
+        return projectKey.equals(that.projectKey) && Objects.equals(branchName, that.branchName);
```
Comment (damien-urruty-sonarsource) [inline src/main/java/org/sonarlint/intellij/core/ProjectBindingManager.java:258]: Interesting, why do we need to keep this overridden `equals` in the record?

**Review by damien-urruty-sonarsource — APPROVED** [review]
Some comments, else LGTM

**Review by eray-felek-sonarsource — APPROVED** [review]

File: clion/src/main/java/org/sonarlint/intellij/clion/AnalyzerConfiguration.java [hunk 2]
```diff
@@ -71,10 +71,9 @@ public ConfigurationResult getConfiguration(VirtualFile file) {
    */
   public ConfigurationResult getConfigurationAction(VirtualFile file) {
     var psiFile = PsiManager.getInstance(project).findFile(file);
-    if (!(psiFile instanceof OCPsiFile)) {
+    if (!(psiFile instanceof OCPsiFile ocFile)) {
       return new ConfigurationResult(psiFile + " not an OCPsiFile");
     }
```
[the remark this answers is not in the document]
Comment (nquinquenel) [reply clion/src/main/java/org/sonarlint/intellij/clion/AnalyzerConfiguration.java:76]: Just below we cast the psi file to and OCPsiFile
`var ocFile = ((OCPsiFile) psiFile).getOCFile();`

File: src/main/java/org/sonarlint/intellij/config/global/ServerConnection.java [hunk 3]
```diff
@@ -87,10 +87,9 @@ private ServerConnection(Builder builder) {
 
   @Override
   public boolean equals(Object o) {
-    if (!(o instanceof ServerConnection)) {
+    if (!(o instanceof ServerConnection other)) {
```
[the remark this answers is not in the document]
Comment (nquinquenel) [reply src/main/java/org/sonarlint/intellij/config/global/ServerConnection.java:90]: Same here, we are casting the object a ServerConnection
`var other = (ServerConnection) o;`

File: src/main/java/org/sonarlint/intellij/core/ProjectBindingManager.java (hunk 1 above)
Comment (nquinquenel) [reply src/main/java/org/sonarlint/intellij/core/ProjectBindingManager.java:258]: Probably for no reason 😄 
The equals [implementation](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Record.html#equals(java.lang.Object)) of the record should be the same

File: clion/src/main/java/org/sonarlint/intellij/clion/AnalyzerConfiguration.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (damien-urruty-sonarsource) [reply clion/src/main/java/org/sonarlint/intellij/clion/AnalyzerConfiguration.java:76]: Yep, in fact I missed that there was a `!` in front, and I didn't know the scope would extend outside of the `if`. Interesting!

File: src/main/java/org/sonarlint/intellij/config/global/SonarLintAboutPanel.java [hunk 4]
```diff
@@ -99,69 +99,70 @@ private JComponent createTelemetryPanel() {
...
+              "hotspot": {
+                "open_in_browser_count": 1,
+                "status_changed_count": 2
+              },
+              "issue": {
+                "status_changed_count": 3
+              },
+              "show_hotspot": {
+                "requests_count": 3
+              },
```
[the remark this answers is not in the document]
Comment (nquinquenel) [reply src/main/java/org/sonarlint/intellij/config/global/SonarLintAboutPanel.java:140]: I added, do you see anything else missing? What's the telemetry for new code period?
```
              "cayc": {
                "new_code_focus": {
                  "enabled": true,
                  "changes": 2
                },
              },
```

[the remark this answers is not in the document]
Comment (nquinquenel) [reply src/main/java/org/sonarlint/intellij/config/global/SonarLintAboutPanel.java:140]: Indeed we also have telemetry for _Open issue in IDE_
```
              "show_issue": {
                "requests_count": 4
              },
```

[the remark this answers is not in the document]
Comment (damien-urruty-sonarsource) [reply src/main/java/org/sonarlint/intellij/config/global/SonarLintAboutPanel.java:146]: Great!

[the remark this answers is not in the document]
Comment (damien-urruty-sonarsource) [reply src/main/java/org/sonarlint/intellij/config/global/SonarLintAboutPanel.java:146]: I said new code period but I was thinking about focus on new code

**Pull request merged**